File: cli/internal/manifest/manifest.go
package manifest

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"os"
	"path/filepath"
	"sort"

	"github.com/anthropics/fastest/cli/internal/ignore"
)

const (
	EntryTypeFile    = "file"
	EntryTypeDir     = "dir"
	EntryTypeSymlink = "symlink"
)

// FileEntry represents a single entry in the manifest
type FileEntry struct {
	Type    string `json:"type"`
	Path    string `json:"path"`
	Hash    string `json:"hash,omitempty"`
	Size    int64  `json:"size,omitempty"`
	Mode    uint32 `json:"mode,omitempty"`
	ModTime int64  `json:"mod_time,omitempty"` // Unix timestamp, optional for reproducibility
	Target  string `json:"target,omitempty"`   // symlink target
}

// Manifest represents a complete project snapshot
type Manifest struct {
	Version string      `json:"version"`
	Files   []FileEntry `json:"files"`
}

// HashFile computes the SHA-256 hash of a file
func HashFile(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}

	return hex.EncodeToString(h.Sum(nil)), nil
}

// Generate creates a manifest for a directory
func Generate(root string, includeModTime bool) (*Manifest, error) {
	// Load ignore patterns
	matcher, err := ignore.LoadFromDir(root)
	if err != nil {
		return nil, err
	}

	manifest := &Manifest{
		Version: "1",
		Files:   []FileEntry{},
	}

	err = filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Get relative path
		relPath, err := filepath.Rel(root, path)
		if err != nil {
			return err
		}

		// Skip root
		if relPath == "." {
			return nil
		}

		// Normalize to forward slashes
		relPath = filepath.ToSlash(relPath)

		// Check if should be ignored
		if matcher.Match(relPath, info.IsDir()) {
			if info.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}

		if info.Mode()&os.ModeSymlink != 0 {
			target, err := os.Readlink(path)
			if err != nil {
				return err
			}
			entry := FileEntry{
				Type:   EntryTypeSymlink,
				Path:   relPath,
				Target: target,
			}
			if includeModTime {
				entry.ModTime = info.ModTime().Unix()
			}
			manifest.Files = append(manifest.Files, entry)
			return nil
		}

		if info.IsDir() {
			entry := FileEntry{
				Type: EntryTypeDir,
				Path: relPath,
				Mode: uint32(info.Mode().Perm()),
			}
			if includeModTime {
				entry.ModTime = info.ModTime().Unix()
			}
			manifest.Files = append(manifest.Files, entry)
			return nil
		}

		// Compute hash
		hash, err := HashFile(path)
		if err != nil {
			return err
		}

		entry := FileEntry{
			Type: EntryTypeFile,
			Path: relPath,
			Hash: hash,
			Size: info.Size(),
			Mode: uint32(info.Mode().Perm()),
		}

		if includeModTime {
			entry.ModTime = info.ModTime().Unix()
		}

		manifest.Files = append(manifest.Files, entry)
		return nil
	})

	if err != nil {
		return nil, err
	}

	// Sort files for reproducibility
	sort.Slice(manifest.Files, func(i, j int) bool {
		if manifest.Files[i].Path == manifest.Files[j].Path {
			return manifest.Files[i].Type < manifest.Files[j].Type
		}
		return manifest.Files[i].Path < manifest.Files[j].Path
	})

	return manifest, nil
}

// ToJSON converts the manifest to canonical JSON
func (m *Manifest) ToJSON() ([]byte, error) {
	return json.MarshalIndent(m, "", "  ")
}

// Hash computes the SHA-256 hash of the manifest
func (m *Manifest) Hash() (string, error) {
	data, err := m.ToJSON()
	if err != nil {
		return "", err
	}

	h := sha256.Sum256(data)
	return hex.EncodeToString(h[:]), nil
}

// FromJSON parses a manifest from JSON
func FromJSON(data []byte) (*Manifest, error) {
	var m Manifest
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, err
	}
	return &m, nil
}

// Diff compares two manifests and returns the differences
func Diff(base, current *Manifest) (added, modified, deleted []string) {
	baseMap := make(map[string]FileEntry)
	for _, f := range base.Files {
		baseMap[f.Path] = f
	}

	currentMap := make(map[string]FileEntry)
	for _, f := range current.Files {
		currentMap[f.Path] = f
	}

	// Find added and modified files
	for _, f := range current.Files {
		if baseFile, exists := baseMap[f.Path]; !exists {
			added = append(added, f.Path)
		} else if !entriesEqual(baseFile, f) {
			modified = append(modified, f.Path)
		}
	}

	// Find deleted files
	for _, f := range base.Files {
		if _, exists := currentMap[f.Path]; !exists {
			deleted = append(deleted, f.Path)
		}
	}

	sort.Strings(added)
	sort.Strings(modified)
	sort.Strings(deleted)

	return added, modified, deleted
}

func entriesEqual(a, b FileEntry) bool {
	if a.Type != b.Type {
		return false
	}
	switch a.Type {
	case EntryTypeFile:
		return a.Hash == b.Hash && a.Mode == b.Mode
	case EntryTypeSymlink:
		return a.Target == b.Target
	case EntryTypeDir:
		return a.Mode == b.Mode
	default:
		return a.Hash == b.Hash && a.Target == b.Target && a.Mode == b.Mode
	}
}

func (m *Manifest) FileEntries() []FileEntry {
	files := make([]FileEntry, 0, len(m.Files))
	for _, f := range m.Files {
		if f.Type == EntryTypeFile {
			files = append(files, f)
		}
	}
	return files
}

func (m *Manifest) DirEntries() []FileEntry {
	dirs := make([]FileEntry, 0, len(m.Files))
	for _, f := range m.Files {
		if f.Type == EntryTypeDir {
			dirs = append(dirs, f)
		}
	}
	return dirs
}

func (m *Manifest) SymlinkEntries() []FileEntry {
	links := make([]FileEntry, 0, len(m.Files))
	for _, f := range m.Files {
		if f.Type == EntryTypeSymlink {
			links = append(links, f)
		}
	}
	return links
}

// TotalSize returns the total size of all files in the manifest
func (m *Manifest) TotalSize() int64 {
	var total int64
	for _, f := range m.Files {
		if f.Type == EntryTypeFile {
			total += f.Size
		}
	}
	return total
}

// FileCount returns the number of files in the manifest
func (m *Manifest) FileCount() int {
	count := 0
	for _, f := range m.Files {
		if f.Type == EntryTypeFile {
			count++
		}
	}
	return count
}

File: packages/shared/src/manifest/manifest.ts
/**
 * Manifest generation and manipulation - matching Go implementation
 */

import type { FileEntry, Manifest, ManifestDiff, FileContent, GenerateOptions } from './types';
import { IgnoreMatcher, DEFAULT_PATTERNS } from './ignore';

/**
 * Compute SHA-256 hash of data
 * Works in both Node.js and browser/Cloudflare Workers environments
 */
export async function sha256(data: Uint8Array | string): Promise<string> {
  let buffer: ArrayBuffer;

  if (typeof data === 'string') {
    buffer = new TextEncoder().encode(data).buffer as ArrayBuffer;
  } else {
    // Handle both ArrayBuffer and Uint8Array
    buffer = data.buffer.slice(data.byteOffset, data.byteOffset + data.byteLength) as ArrayBuffer;
  }

  const hashBuffer = await crypto.subtle.digest('SHA-256', buffer);
  const hashArray = new Uint8Array(hashBuffer);

  return Array.from(hashArray)
    .map(b => b.toString(16).padStart(2, '0'))
    .join('');
}

/**
 * Generate a manifest from a list of files with their content
 * This is the main function for cloud-side manifest generation
 */
export async function generateFromFiles(
  files: FileContent[],
  options: GenerateOptions = {}
): Promise<Manifest> {
  const { includeModTime = false, ignorePatterns } = options;

  // Create matcher with custom patterns if provided
  const patterns = ignorePatterns
    ? [...DEFAULT_PATTERNS, ...ignorePatterns]
    : DEFAULT_PATTERNS;
  const matcher = new IgnoreMatcher(patterns);

  const entries: FileEntry[] = [];

  for (const file of files) {
    // Normalize path
    const path = file.path.replace(/\\/g, '/');

    // Skip ignored files
    if (matcher.match(path, false)) {
      continue;
    }

    // Compute hash
    const content = typeof file.content === 'string'
      ? new TextEncoder().encode(file.content)
      : file.content;
    const hash = await sha256(content);

    const entry: FileEntry = {
      path,
      hash,
      size: content.length,
      mode: file.mode ?? 0o644, // Default to readable file
    };

    if (includeModTime && file.modTime !== undefined) {
      entry.mod_time = file.modTime;
    }

    entries.push(entry);
  }

  // Sort files for reproducibility (matching Go implementation)
  entries.sort((a, b) => a.path.localeCompare(b.path));

  return {
    version: '1',
    files: entries,
  };
}

/**
 * Convert manifest to canonical JSON (matching Go's json.MarshalIndent)
 */
export function toJSON(manifest: Manifest): string {
  return JSON.stringify(manifest, null, '  ');
}

/**
 * Compute the SHA-256 hash of a manifest
 */
export async function hashManifest(manifest: Manifest): Promise<string> {
  const json = toJSON(manifest);
  return sha256(json);
}

/**
 * Parse a manifest from JSON
 */
export function fromJSON(data: string): Manifest {
  const parsed = JSON.parse(data);

  // Validate basic structure
  if (typeof parsed.version !== 'string' || !Array.isArray(parsed.files)) {
    throw new Error('Invalid manifest format');
  }

  return parsed as Manifest;
}

/**
 * Compare two manifests and return the differences
 */
export function diff(base: Manifest, current: Manifest): ManifestDiff {
  const baseMap = new Map<string, FileEntry>();
  for (const f of base.files) {
    baseMap.set(f.path, f);
  }

  const currentMap = new Map<string, FileEntry>();
  for (const f of current.files) {
    currentMap.set(f.path, f);
  }

  const added: string[] = [];
  const modified: string[] = [];
  const deleted: string[] = [];

  // Find added and modified files
  for (const f of current.files) {
    const baseFile = baseMap.get(f.path);
    if (!baseFile) {
      added.push(f.path);
    } else if (baseFile.hash !== f.hash) {
      modified.push(f.path);
    }
  }

  // Find deleted files
  for (const f of base.files) {
    if (!currentMap.has(f.path)) {
      deleted.push(f.path);
    }
  }

  // Sort for consistent output
  added.sort();
  modified.sort();
  deleted.sort();

  return { added, modified, deleted };
}

/**
 * Get total size of all files in the manifest
 */
export function totalSize(manifest: Manifest): number {
  return manifest.files.reduce((sum, f) => sum + f.size, 0);
}

/**
 * Get the number of files in the manifest
 */
export function fileCount(manifest: Manifest): number {
  return manifest.files.length;
}

/**
 * Get a file entry by path
 */
export function getFile(manifest: Manifest, path: string): FileEntry | undefined {
  return manifest.files.find(f => f.path === path);
}

/**
 * Get all unique blob hashes from a manifest
 */
export function getBlobHashes(manifest: Manifest): string[] {
  const hashes = new Set<string>();
  for (const f of manifest.files) {
    hashes.add(f.hash);
  }
  return Array.from(hashes).sort();
}

/**
 * Find blobs that exist in current but not in base (for incremental upload)
 */
export function getNewBlobHashes(base: Manifest, current: Manifest): string[] {
  const baseHashes = new Set(base.files.map(f => f.hash));
  const newHashes = new Set<string>();

  for (const f of current.files) {
    if (!baseHashes.has(f.hash)) {
      newHashes.add(f.hash);
    }
  }

  return Array.from(newHashes).sort();
}

/**
 * Create an empty manifest
 */
export function empty(): Manifest {
  return {
    version: '1',
    files: [],
  };
}
